socorro/cron/crontabber_app.py

```diff
@@ -3,67 +3,109 @@
+]
+
+DEFAULT_JOBS = ', '.join(DEFAULT_JOBS_BASE)
+STAGE_JOBS = ', '.join(
+    DEFAULT_JOBS_BASE +
+    ['socorro.cron.jobs.fetch_adi_from_hive.FAKEFetchADIFromHiveCronApp|1d']
+)
+
+
+def jobs_converter(path_or_jobs):
+    """Takes a Python dotted path or a jobs spec and returns crontabber jobs
+
+    Example Python dotted path::
+
+        jobs_converter('socorro.cron.crontabber_app.DEFAULT_JOBS')
+
+    Example jobs spec::
+
+        jobs_converter('socorro.cron.jobs.ftpscraper.FTPScraperCronApp|1h')
+
+
+    :arg str path_or_jobs: a Python dotted path or a crontabber jobs spec
+
+    :returns: crontabber jobs InnerClassList
+
+    """
+    if '|' not in path_or_jobs:
+        # NOTE(willkg): crontabber's class_converter returns the value pointed
+        # to by a Python dotted path
+        input_str = class_converter(path_or_jobs)
+    else:
+        input_str = path_or_jobs
+
+    from_string_converter = classes_in_namespaces_converter_with_compression(
+        reference_namespace=Namespace(),
+        list_splitter_fn=line_splitter,
+        class_extractor=pipe_splitter,
+        extra_extractor=get_extra_as_options
+    )
+    return from_string_converter(input_str)
 
 
 class CronTabberApp(CronTabberBase, App):
-    @staticmethod
-    def get_application_defaults():
-        return {
-            'jobs': DEFAULT_JOBS,
-            'crontabber.database_class':
-            'socorro.external.postgresql.connection_context.ConnectionContext',
-            'crontabber.transaction_executor_class':
-            'socorro.database.transaction_executor.TransactionExecutor',
-            'crontabber.job_state_db_class.database_class':
-            'socorro.external.postgresql.connection_context.ConnectionContext',
-
-        }
-
-
-# These settings should ideally be done in config, but because, at
-# the moment, it's easier for us to maintain python we're doing it here.
-CronTabber.required_config.crontabber.jobs.default = DEFAULT_JOBS
-CronTabber.required_config.crontabber.database_class.default = (
+    required_config = CronTabberBase.required_config.safe_copy()
+
+    # Stomp on the existing "crontabber.jobs" configuration with one that can
+    # also take Python dotted paths
+    required_config.crontabber.add_option(
+        'jobs',
+        default='socorro.cron.crontabber_app.DEFAULT_JOBS',
+        from_string_converter=jobs_converter,
+        doc='Crontabber jobs spec or Python dotted path to jobs spec',
+    )
+
+
+# NOTE(willkg): We need to "fix" the defaults here rather than use App's
+# get_application_defaults() because that doesn't support nested configuration
+# defaults
+CronTabberApp.required_config.crontabber.database_class.default = (
+    'socorro.external.postgresql.connection_context.ConnectionContext'
+)
+CronTabberApp.required_config.crontabber.job_state_db_class.default.required_config.database_class.default = (  # noqa
     'socorro.external.postgresql.connection_context.ConnectionContext'
 )
```

Comment:
This is so messy, but I couldn't figure out how to fix `get_application_defaults` in a way that didn't affect a ton of other things.

Reply:
`        reference_value_from='resource.postgresql',` ?

Reply:
The defaults here are coming from `CronTabber` which is from the crontabber library. We could change the `default` or `reference_value_from` for `database_class` by stomping on it in `CronTabberApp`, but in order to change the default for the `database_class` for `job_state_db_class`, we'd have to subclass `job_state_db_class` and stomp on the option there and change the default or add a `reference_value_from` and then use our subclass. That's doable, but adds a bunch more code.

In the case here, we're only using crontabber one way so (ab)using the classes seems like a better choice.

I'm pretty sure `get_application_defaults` is supposed to fix this issue by letting you adjust defaults to a more application-appropriate value.

https://github.com/mozilla-services/socorro/blob/master/socorro/app/socorro_app.py#L147

But it doesn't seem to work with nested items and I didn't want to add fiddling with that to this PR.